### PR TITLE
Improve Docker image

### PR DESCRIPTION
**Pull request by kanolato**
When im using Docker i like that i dont have to install anything else but Docker. I thought that it would be nice to build the project in `gatsbyjs/gatsby:onbuild` and then just copy the public dir to `gatsbyjs/gatsby`. Now as the documentation says the `:onbuild` tag builds the project and `:latest` serves it.
Also i first tried to use the image as the documentation but the `:onbuild` tag lacks of nginx rules and every request gets a `500`.
#25

### CODE REVIEW

**Comment by polarathene** [comment]
Ok. So these changes create this flow:

- Switch from the `alpine:edge` base to a Debian Node base image.
- Install a fixed version of gatsby cli **(why is the version fixed?)**
- Copy over the project into the image
- Install via yarn followed by a gatsby build
- Require the user to add two files as described in the documention
- Have the user build that local dockerfile, performing the referenced `ONBUILD` instructions
- Followed by a multi-stage build to copy the build output in public to the `alpine:edge` base image

Why was the port changed from `80` to `8080`? Your related commit says due to "practicity" What does that mean?

This change while nice, will prevent existing users of the current workflow from being able to avoid the build within the container and debugging anything that may go wrong. Perhaps this contribution should rather compliment the existing offering?

This doesn't solve #25, as that's about a development image. One that doesn't require building an image locally for every change.

> Also i first tried to use the image as the documentation but the :onbuild tag lacks of nginx rules and every request gets a 500.

What has your PR done to resolve that? Both end up serving nginx from the same main gatsby alpine image?

File: Dockerfile.onbuild [hunk 1]
```diff
@@ -1,4 +1,7 @@
-FROM gatsbyjs/gatsby:latest
+FROM node:10.15.3-stretch
```
Comment (polarathene) [inline Dockerfile.onbuild]: Why this specific version? Is `node:10-stretch` not sufficient? 

This tag is a bit overly specific and would not receive newer updates to the Node 10 image.

File: Dockerfile.onbuild [hunk 2]
```diff
@@ -1,4 +1,7 @@
...
 MAINTAINER José Moreira <josemoreiravarzim@gmail.com>
-
-ONBUILD ADD public/ /pub
+RUN yarn global add gatsby-cli@2.4.17
```
Comment (polarathene) [inline Dockerfile.onbuild]: Any reason for locking to `@2.4.17`?

This will prevent a user from benefiting from future improvements to the CLI wouldn't?

File: README.md [hunk 3]
```diff
@@ -26,8 +36,8 @@ This image has two major tags:
     ```
 1. Use it
     ```bash
-    docker run --rm -p 80:80 myproject/website
-    # Open your browser at http://localhost
+    docker run --rm -p 8080:80 myproject/website
```
Comment (polarathene) [inline README.md]: Why the change to port 8080? 

Would it be better to explain in the documentation more clearly?

**Comment by JaKXz** [comment]
@kanolato I am in agreement with @polarathene's comments - are you able to update this PR and maybe it could get merged?

**Comment by kanolato** [comment]
Going to update it in the following days :) @JaKXz

File: Dockerfile.onbuild (hunk 1 above)
Comment (cusspvz) [reply Dockerfile.onbuild]: isn't better to use alpine instead?

File: README.md [hunk 4]
```diff
@@ -11,11 +11,21 @@ This image has two major tags:
...
+    _Dockerfile_
     ```dockerfile
-    FROM gatsbyjs/gatsby:onbuild
+    FROM gatsbyjs/gatsby:onbuild as build
+
+    FROM gatsbyjs/gatsby
+    COPY --from=build /app/public /pub
+    ```
+    _.dockerignore_
+    ```dockerfile
```
Comment (cusspvz) [inline README.md]:
```diff
-```dockerfile
+```ignore
```

File: Dockerfile.onbuild (hunk 1 above)
Comment (polarathene) [reply Dockerfile.onbuild]: @cusspvz for production, but as a generic development image not really. I had already made a Docker image with Alpine for Gatsby but it was problematic because of the image packages not having good compatibility.

File: Dockerfile.onbuild [hunk 5]
```diff
@@ -1,4 +1,7 @@
...
+ONBUILD WORKDIR /app
+ONBUILD ADD . ./
+ONBUILD RUN yarn
+ONBUILD RUN gatsby build
```
Comment (cdaringe) [inline Dockerfile.onbuild]: for most gatsby users, i suspect this new file will be missing required system dependencies for native compile steps from core-js|cwebp|sharp|etc used in the gatsby build

**Comment by kanolato** [comment]
Than you all for your comments!. The required changes were just pushed.

File: Dockerfile.onbuild (hunk 5 above)
Comment (kanolato) [reply Dockerfile.onbuild]: Sorry I'm not getting it, wich new file are you referring to?

Comment (kanolato) [reply Dockerfile.onbuild]: you mean this deps should also be added to the image? `core-js|cwebp|sharp`

**Comment by kanolato** [comment]
> #25

My bad, You are completely right on the last part, this doesn't solve #25, as in the end it only provides a way to serve the built assets. I will explore it later next week. Im almost sure its possible mounting volumes and so. Again thank you for your replies

**Comment by polarathene** [comment]
IIRC, stretch image dependencies were out of date, especially cwebp. Since Buster has been out for a while, perhaps better to switch to that.

**Comment by kanolato** [comment]
I have changed to buster and upgrade node to 12

File: Dockerfile.onbuild (hunk 5 above)
Comment (cdaringe) [reply Dockerfile.onbuild]: correct.  those things have native compile toolchains, of which the support libraries required for such installations i suspect are _not present_ in this image.  gatsby will eventually try and use node-native bindings to hook into them on build, but would not be found.  thus, i suggest `gatsby build` would fail
